File: LibretroRTSupport/Converter.cpp
#include "pch.h"
#include "Converter.h"
#include "../LibretroRT/libretro.h"

Platform::String^ LibretroRTSupport::Converter::CToPlatformString(const char* t_str)
{
	//setup converter
	typedef std::codecvt_utf8<wchar_t> convert_type;
	std::wstring_convert<convert_type, wchar_t> converter;

	//use converter (.to_bytes: wstr->str, .from_bytes: str->wstr)
	auto wstring = converter.from_bytes(t_str);
	return ref new Platform::String(wstring.c_str());
}

LibretroRT::GameGeometry^ LibretroRTSupport::Converter::CToRTGameGeometry(const retro_game_geometry & geometry)
{
	return ref new LibretroRT::GameGeometry(geometry.base_width, geometry.base_height, geometry.max_width, geometry.max_height, geometry.aspect_ratio);
}

LibretroRT::SystemTiming ^ LibretroRTSupport::Converter::CToRTSystemTiming(const retro_system_timing & timing)
{
	return ref new LibretroRT::SystemTiming(timing.fps, timing.sample_rate);
}

InputType LibretroRTSupport::Converter::ConvertToInputType(unsigned device, unsigned index, unsigned id)
{
	switch (device)
	{
	case RETRO_DEVICE_JOYPAD:
		switch (id)
		{
		case RETRO_DEVICE_ID_JOYPAD_B:
			return InputType::DeviceIdJoypadB;
		case RETRO_DEVICE_ID_JOYPAD_Y:
			return InputType::DeviceIdJoypadY;
		case RETRO_DEVICE_ID_JOYPAD_SELECT:
			return InputType::DeviceIdJoypadSelect;
		case RETRO_DEVICE_ID_JOYPAD_START:
			return InputType::DeviceIdJoypadStart;
		case RETRO_DEVICE_ID_JOYPAD_UP:
			return InputType::DeviceIdJoypadUp;
		case RETRO_DEVICE_ID_JOYPAD_DOWN:
			return InputType::DeviceIdJoypadDown;
		case RETRO_DEVICE_ID_JOYPAD_LEFT:
			return InputType::DeviceIdJoypadLeft;
		case RETRO_DEVICE_ID_JOYPAD_RIGHT:
			return InputType::DeviceIdJoypadRight;
		case RETRO_DEVICE_ID_JOYPAD_A:
			return InputType::DeviceIdJoypadA;
		case RETRO_DEVICE_ID_JOYPAD_X:
			return InputType::DeviceIdJoypadX;
		case RETRO_DEVICE_ID_JOYPAD_L:
			return InputType::DeviceIdJoypadL;
		case RETRO_DEVICE_ID_JOYPAD_R:
			return InputType::DeviceIdJoypadR;
		case RETRO_DEVICE_ID_JOYPAD_L2:
			return InputType::DeviceIdJoypadL2;
		case RETRO_DEVICE_ID_JOYPAD_R2:
			return InputType::DeviceIdJoypadR2;
		case RETRO_DEVICE_ID_JOYPAD_L3:
			return InputType::DeviceIdJoypadL3;
		case RETRO_DEVICE_ID_JOYPAD_R3:
			return InputType::DeviceIdJoypadR3;
		default:
			return InputType::DeviceIdUnknown;
		}
	case RETRO_DEVICE_ANALOG:
		switch (id)
		{
		default:
			return InputType::DeviceIdUnknown;
		}
	case RETRO_DEVICE_MOUSE:
		switch (id)
		{
		default:
			return InputType::DeviceIdUnknown;
		}
	case RETRO_DEVICE_LIGHTGUN:
		switch (id)
		{
		default:
			return InputType::DeviceIdUnknown;
		}
	case RETRO_DEVICE_POINTER:
		switch (id)
		{
		default:
			return InputType::DeviceIdUnknown;
		}
	default:
		return InputType::DeviceIdUnknown;
	}
}

File: LibRetriX/LibRetriX.RetroBindings/Tools/Converter.cs
namespace LibRetriX.RetroBindings.Tools
{
    public static class Converter
    {
        public static InputTypes ConvertToInputType(uint device, uint index, uint id)
        {
            switch (device)
            {
                case Constants.RETRO_DEVICE_JOYPAD:
                    switch (id)
                    {
                        case Constants.RETRO_DEVICE_ID_JOYPAD_B:
                            return InputTypes.DeviceIdJoypadB;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_Y:
                            return InputTypes.DeviceIdJoypadY;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_SELECT:
                            return InputTypes.DeviceIdJoypadSelect;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_START:
                            return InputTypes.DeviceIdJoypadStart;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_UP:
                            return InputTypes.DeviceIdJoypadUp;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_DOWN:
                            return InputTypes.DeviceIdJoypadDown;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_LEFT:
                            return InputTypes.DeviceIdJoypadLeft;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_RIGHT:
                            return InputTypes.DeviceIdJoypadRight;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_A:
                            return InputTypes.DeviceIdJoypadA;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_X:
                            return InputTypes.DeviceIdJoypadX;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_L:
                            return InputTypes.DeviceIdJoypadL;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_R:
                            return InputTypes.DeviceIdJoypadR;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_L2:
                            return InputTypes.DeviceIdJoypadL2;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_R2:
                            return InputTypes.DeviceIdJoypadR2;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_L3:
                            return InputTypes.DeviceIdJoypadL3;
                        case Constants.RETRO_DEVICE_ID_JOYPAD_R3:
                            return InputTypes.DeviceIdJoypadR3;
                        default:
                            return InputTypes.DeviceIdUnknown;
                    }
                case Constants.RETRO_DEVICE_ANALOG:
                    switch (index)
                    {
                        case Constants.RETRO_DEVICE_INDEX_ANALOG_LEFT:
                            switch (id)
                            {
                                case Constants.RETRO_DEVICE_ID_ANALOG_X:
                                    return InputTypes.DeviceIdAnalogLeftX;
                                case Constants.RETRO_DEVICE_ID_ANALOG_Y:
                                    return InputTypes.DeviceIdAnalogLeftY;
                                default:
                                    return InputTypes.DeviceIdUnknown;
                            }
                        case Constants.RETRO_DEVICE_INDEX_ANALOG_RIGHT:
                            switch (id)
                            {
                                case Constants.RETRO_DEVICE_ID_ANALOG_X:
                                    return InputTypes.DeviceIdAnalogRightX;
                                case Constants.RETRO_DEVICE_ID_ANALOG_Y:
                                    return InputTypes.DeviceIdAnalogRightY;
                                default:
                                    return InputTypes.DeviceIdUnknown;
                            }
                        default:
                            return InputTypes.DeviceIdUnknown;
                    }
                case Constants.RETRO_DEVICE_MOUSE:
                    switch (id)
                    {
                        case Constants.RETRO_DEVICE_ID_MOUSE_X:
                            return InputTypes.DeviceIdMouseX;
                        case Constants.RETRO_DEVICE_ID_MOUSE_Y:
                            return InputTypes.DeviceIdMouseY;
                        case Constants.RETRO_DEVICE_ID_MOUSE_LEFT:
                            return InputTypes.DeviceIdMouseLeft;
                        case Constants.RETRO_DEVICE_ID_MOUSE_RIGHT:
                            return InputTypes.DeviceIdMouseRight;
                        case Constants.RETRO_DEVICE_ID_MOUSE_WHEELUP:
                            return InputTypes.DeviceIdMouseWheelup;
                        case Constants.RETRO_DEVICE_ID_MOUSE_WHEELDOWN:
                            return InputTypes.DeviceIdMouseWheeldown;
                        case Constants.RETRO_DEVICE_ID_MOUSE_MIDDLE:
                            return InputTypes.DeviceIdMouseMiddle;
                        case Constants.RETRO_DEVICE_ID_MOUSE_HORIZ_WHEELUP:
                            return InputTypes.DeviceIdMouseHorizWheelup;
                        case Constants.RETRO_DEVICE_ID_MOUSE_HORIZ_WHEELDOWN:
                            return InputTypes.DeviceIdMouseHorizWheeldown;
                        default:
                            return InputTypes.DeviceIdUnknown;
                    }
                case Constants.RETRO_DEVICE_LIGHTGUN:
                    switch (id)
                    {
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_X:
                            return InputTypes.DeviceIdLightgunX;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_Y:
                            return InputTypes.DeviceIdLightgunY;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_TRIGGER:
                            return InputTypes.DeviceIdLightgunTrigger;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_CURSOR:
                            return InputTypes.DeviceIdLightgunCursor;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_TURBO:
                            return InputTypes.DeviceIdLightgunTurbo;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_PAUSE:
                            return InputTypes.DeviceIdLightgunPause;
                        case Constants.RETRO_DEVICE_ID_LIGHTGUN_START:
                            return InputTypes.DeviceIdLightgunStart;
                        default:
                            return InputTypes.DeviceIdUnknown;
                    }
                case Constants.RETRO_DEVICE_POINTER:
                    switch (id)
                    {
                        case Constants.RETRO_DEVICE_ID_POINTER_X:
                            return InputTypes.DeviceIdPointerX;
                        case Constants.RETRO_DEVICE_ID_POINTER_Y:
                            return InputTypes.DeviceIdPointerY;
                        case Constants.RETRO_DEVICE_ID_POINTER_PRESSED:
                            return InputTypes.DeviceIdPointerPressed;
                        default:
                            return InputTypes.DeviceIdUnknown;
                    }
                default:
                    return InputTypes.DeviceIdUnknown;
            }
        }
    }
}
